generators/python/core_utilities/sdk/pagination.py

```diff
@@ -0,0 +1,71 @@
+import typing
+from typing_extensions import Self
+from .pydantic_utilities import pydantic_v1
+import asyncio
+
+EMPTY_PROMISE = asyncio.sleep(0)
+
+# Generic to represent the underlying type of the results within a page
+T = typing.TypeVar("T")
+
+# SDKs implement a Page ABC per-pagination request, the endpoint then retuns a paginator that wraps this type
+# for example, an endpoint will return SyncPaginator[UserPage] where UserPage implements the Page ABC. ex:
+# 
+# SyncPaginator<InnerListType>(
+#     has_next=response.list_metadata.after is not None,
+#     items=response.data,
+#     # This should be the outer function that returns the SyncPaginator again
+#     get_next=lambda: list(..., cursor: response.cursor) (or list(..., offset: offset + 1))
+# )
+class BasePage(pydantic_v1.BaseModel, typing.Generic[T]):
+    has_next: bool
+    items: typing.List[T]
+
+class SyncPage(BasePage, typing.Generic[T]):
+    get_next: typing.Optional[typing.Callable[[], typing.Optional[Self]]]
+
+class AsyncPage(BasePage, typing.Generic[T]):
+    get_next: typing.Optional[typing.Callable[[], typing.Awaitable[typing.Optional[Self]]]]
+
+# ----------------------------
+
+class SyncPaginator(SyncPage[T], typing.Generic[T]):
```

Comment:
Should we align the naming here with Go like `SyncPager`, `AsyncPager` ?

Reply:
Updating!